src/engine/loopingcontrol.cpp

```diff
@@ -711,151 +820,280 @@ void LoopingControl::clearActiveBeatLoop() {
+        return;
+    }
+
+    // O(n) search, but there are only ~10-ish beatloop controls so this is
+    // fine.
+    for (BeatLoopingControl* pBeatLoopControl: m_beatLoops) {
+        if (pBeatLoopControl->getSize() == beats) {
+            if (enable || m_bLoopingEnabled) {
+                pBeatLoopControl->activate();
+            }
+            BeatLoopingControl* pOldBeatLoop =
+                    m_pActiveBeatLoop.fetchAndStoreRelease(pBeatLoopControl);
+            if (pOldBeatLoop != nullptr && pOldBeatLoop != pBeatLoopControl) {
+                pOldBeatLoop->deactivate();
+            }
+            break;
+        }
+    }
+
+    // If resizing an inactive loop by changing beatloop_size,
+    // do not seek to the adjusted loop.
+    if (keepStartPoint && (enable || m_bLoopingEnabled)) {
         seekInsideAdjustedLoop(loopSamples.start, loopSamples.end,
                 newloopSamples.start, newloopSamples.end);
     }
 
     m_loopSamples.setValue(newloopSamples);
     m_pCOLoopStartPosition->set(newloopSamples.start);
     m_pCOLoopEndPosition->set(newloopSamples.end);
-    setLoopingEnabled(true);
+    if (enable) {
+        setLoopingEnabled(true);
+    }
+    m_bLoopSetSinceTrackLoaded = true;
+}
+
+void LoopingControl::slotBeatLoopSizeChangeRequest(double beats) {
+    // slotBeatLoop will call m_pCOBeatLoopSize->setAndConfirm if
+    // new beatloop_size is valid
+    slotBeatLoop(beats, true, false);
+}
+
+void LoopingControl::slotBeatLoopToggle(double pressed) {
+    if (pressed > 0) {
+        if (m_bLoopingEnabled) {
+            setLoopingEnabled(false);
+        } else {
+            slotBeatLoop(m_pCOBeatLoopSize->get());
+        }
+    }
+}
+
+void LoopingControl::slotBeatLoopRollToggle(double pressed) {
+    if (pressed > 0.0) {
+        if (m_bLoopingEnabled) {
+            setLoopingEnabled(false);
+            // Make sure slip mode is not turned off if it was turned on
+            // by something that was not a rolling beatloop.
+            if (m_bLoopRollActive) {
+                m_pSlipEnabled->set(0.0);
+                m_bLoopRollActive = false;
+            }
+        } else {
+            m_pSlipEnabled->set(1.0);
+            slotBeatLoop(m_pCOBeatLoopSize->get());
+            m_bLoopRollActive = true;
+        }
+    } else {
+        setLoopingEnabled(false);
+        // Make sure slip mode is not turned off if it was turned on
+        // by something that was not a rolling beatloop.
+        if (m_bLoopRollActive) {
+            m_pSlipEnabled->set(0.0);
+            m_bLoopRollActive = false;
+        }
+    }
+}
+
+void LoopingControl::slotLoopAutoToggle(double pressed) {
+    double beatloop_size = m_pCOBeatLoopSize->get();
+    if (beatloop_size > 1.0) {
+        slotBeatLoopToggle(pressed);
+    } else {
+        slotBeatLoopRollToggle(pressed);
+    }
+}
+
+void LoopingControl::slotLoopManualToggle(double pressed) {
+    if (m_bLoopingEnabled) {
```

Comment:
I think the code is pretty self-explanatory here, but I could add a comment:
If a loop is enabled, disable it. Otherwise, set the loop in point on button down and loop out on button up.

I could call `setLoopingEnabled(false)` instead of `slotReloopToggle`. Would that be clearer?